### DISCUSSION TITLE
Where does Gramine source its entropy for RA-TLS private key generation?

### DISCUSSION THREAD

**Discussion by fwoodruff** [gramineproject/gramine#1667] [Q&A] [answered]
Could someone MITM between Linux urandom and Gramine's private key generation function? Or does Gramine use an SGX primitive for entropy generation?

**Comment by dimakuv** [comment, marked as answer]
Gramine uses the RDRAND instruction as the sole source of entropy.

Similarly, RA-TLS (which is not a part of Gramine core per se, but a library that is used together with Gramine) also uses the RDRAND instruction as the sole source of entropy.

Since you're asking about RA-TLS, here are the RA-TLS specific code points:
- https://github.com/gramineproject/gramine/blob/fa437cbcf380ea9c1aa76958ddeb73557a344901/subprojects/packagefiles/mbedtls/include/mbedtls/config-pal.h#L20 -- RA-TLS uses our own patched version of mbedTLS under the hood (linked statically), which has the `MBEDTLS_ENTROPY_HARDWARE_ALT` macro that forces mbedTLS to use our own provided `mbedtls_hardware_poll()` function
- https://github.com/gramineproject/gramine/blob/fa437cbcf380ea9c1aa76958ddeb73557a344901/common/src/crypto/adapters/mbedtls_adapter.c#L277-L290 -- our implementation of `mbedtls_hardware_poll()`, which calls `PalRandomBitsRead()`
- https://github.com/gramineproject/gramine/blob/fa437cbcf380ea9c1aa76958ddeb73557a344901/pal/src/host/linux-sgx/pal_misc.c#L779-L786 -- `PalRandomBitsRead()` ends up here, which calls `rdrand()`
- https://github.com/gramineproject/gramine/blob/fa437cbcf380ea9c1aa76958ddeb73557a344901/common/include/arch/x86_64/cpu.h#L75-L83 -- `rdrand()` is a wrapper around RDRAND instruction

> Could someone MITM between Linux urandom and Gramine's private key generation function?

As can be seen from the above code, no, Gramine and RA-TLS do **not** use Linux's unrandom. We rely purely on the HW primitives.

**Reply by dimakuv** [reply]
Before anyone else asks, here's how Gramine emulates the `/dev/random` and `/dev/urandom` pseudo-files for the apps:
- https://github.com/gramineproject/gramine/blob/fa437cbcf380ea9c1aa76958ddeb73557a344901/libos/src/fs/dev/fs.c#L138-L142
- https://github.com/gramineproject/gramine/blob/fa437cbcf380ea9c1aa76958ddeb73557a344901/libos/src/fs/dev/fs.c#L147-L152
- https://github.com/gramineproject/gramine/blob/fa437cbcf380ea9c1aa76958ddeb73557a344901/libos/src/fs/dev/fs.c#L50-L57

So similarly to the RA-TLS case, Gramine core itself uses (and exposes to the app) the interface that uses RDRAND instruction only (a pure HW primitive).
